[feat] add favicons to result urls (#3727)
metasearch-lab/searx: docs/admin/searx.favicons.rst

and not well suitable for a productive environment!

Setting up the cache
====================

To parameterize the *cache* and more settings of the favicons infrastructure, a
TOML_ configuration is created in the file ``/etc/searxng/favicons.toml``.

.. _TOML: https://toml.io/en/

.. code:: toml

   [favicons]

   cfg_schema = 1   # config's schema version no.

   [favicons.cache]

   db_url = "/var/cache/searxng/faviconcache.db"  # default: "/tmp/faviconcache.db"
   LIMIT_TOTAL_BYTES = 2147483648                 # 2 GB / default: 50 MB
   # HOLD_TIME = 5184000                            # 60 days / default: 30 days
   # BLOB_MAX_BYTES = 40960                         # 40 KB / default 20 KB
   # MAINTENANCE_MODE = "off"                       # default: "auto"
   # MAINTENANCE_PERIOD = 600                       # 10min / default: 1h

:py:obj:`cfg_schema <.FaviconConfig.cfg_schema>`:
  Is required to trigger any processes required for future upgrades / don't
  change it.

:py:obj:`cache.db_url <.FaviconCacheConfig.db_url>`:
  The path to the (SQLite_) database file.  The default path is in the `/tmp`_
  folder, which is deleted on every reboot and is therefore unsuitable for a

>   productive environment.  The FHS_ provides the folder  for the

Author: 'productive environment' -> 'production environment'

.. _favicons:

========
Favicons
========

.. sidebar:: warning

>    Don't activate the favacions before reading the documentation.

Author: favacions -> favicons

.. contents::
   :depth: 2
   :local:
   :backlinks: entry

Activating the favicons in SearXNG is very easy, but this **generates a

> significantly higher load** in the communication between client and server and
> additional resources are used on the server.

Reviewer: ```suggestion
significantly higher load** in the communication between client and server and
increases resource usage on the server.
```
nit: two `and`s, grammar

Reviewer 2: my intention was to say: *"higher load in the client/server communication and increases resource usage on the server."* --> [fixed](https://github.com/searxng/searxng/compare/a76b37fe75f50c8281fbec24499e74211ba2b67e..7c7482fc107484e883803df65176647d58aff989)

To mitigate these disadvantages, various methods have been implemented,
including a *cache*.  The cache must be parameterized according to your own
requirements and maintained regularly.

> To activate favicons in SearXNG's result list, set a default favicons-resolver

Reviewer: ```suggestion
To activate favicons in SearXNG's result list, set a default ``favicon_resolver``
```
named favicon_resolver without s below

> To activate favicons in SearXNG's result list set a default favicons-resolver in

Author: Add comma:

To activate favicons in SearXNG's result list set a default favicons-resolver in

To activate favicons in SearXNG's result list, set a default favicons-resolver in

the :ref:`search <settings search>` settings:

.. code:: yaml

   search:
     favicon_resolver: "duckduckgo"

> By default and without any extensions SearXNG serves these resolvers:

Author: Add comma:

By default and without any extensions SearXNG serves these resolvers:

By default and without any extensions, SearXNG serves these resolvers:

- ``duckduckgo``
- ``allesedv``
- ``google``
- ``yandex``

> With the above setting favicons are displayed, the user has the option to
> deactivate this feature in his settings.  If the user is to have the option of
> selecting from several *resolvers*, a further setting is required / *but this will
> be discussed later, first we have to setup the favicons cache*.

Reviewer: ```suggestion
With the above setting, favicons are displayed by default. Users have the option to
deactivate this feature in their settings..
```
Forgot to remove? / TODO?

Reviewer 2: Nope .. "but this will be discussed later" means later in this article .. but I [fixed](https://github.com/searxng/searxng/compare/a76b37fe75f50c8281fbec24499e74211ba2b67e..7c7482fc107484e883803df65176647d58aff989) this part to be more clear.

Infrastructure
==============

The infrastructure for providing the favicons essentially consists of three
parts:

- :py:obj:`Favicons-Proxy <.favicons.proxy>` (aka *proxy*)

> - :py:obj:`Favicons-Resolvers <.favicons.resolvers>` (aka *reolver*)

Author: reolver -> resolver

   # MAINTENANCE_PERIOD = 600                       # 10min / default: 1h

:py:obj:`cfg_schema <.FaviconConfig.cfg_schema>`:
  Is required to trigger any processes required for future upgrades / don't
  change it.

:py:obj:`cache.db_url <.FaviconCacheConfig.db_url>`:
  The path to the (SQLite_) database file.  The default path is in the `/tmp`_
  folder, which is deleted on every reboot and is therefore unsuitable for a
  production environment.  The FHS_ provides the folder  for the
  application cache

  The FHS_ provides the folder `/var/cache`_ for the cache of applications, so a
  suitable storage location of SearXNG's caches is folder ``/var/cache/searxng``.
  In container systems, a volume should be mounted for this folder and in a
  standard installation (compare :ref:`create searxng user`), the folder must be
  created and the user under which the SearXNG process is running must be given
  write permission to this folder.

  .. code:: bash

     $ sudo mkdir /var/cache/searxng
     $ sudo chown root:searxng /var/cache/searxng/
     $ sudo chmod g+w /var/cache/searxng/

:py:obj:`cache.LIMIT_TOTAL_BYTES <.FaviconCacheConfig.LIMIT_TOTAL_BYTES>`:

>   Maximum of bytes (default) stored in the cache of all blobs. The limit is only

Author: Maximum of bytes (default)?

Did you forget to include the figure?
